### ISSUE TITLE
NullPointerException after 3 connection exceptions using FallbackHostHandler

### ISSUE THREAD

**Issue by hanyu-natsu** [arangodb/arangodb-java-driver#246]
version: 5.0.3
java version: 1.8
with arangodb-spring-data: 3.2.1

arangodb driver fails to execute requests because of NullPointerException.
"host" seems to be null in some case but not checked in execution section.
However it does have null check in the exception handling.

```java
java.lang.NullPointerException: null
        at com.arangodb.internal.http.HttpCommunication.execute(HttpCommunication.java:87) ~[arangodb-java-driver-5.0.3.jar!/:?]
        at com.arangodb.internal.http.HttpProtocol.execute(HttpProtocol.java:49) ~[arangodb-java-driver-5.0.3.jar!/:?]
        at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:71) ~[arangodb-java-driver-5.0.3.jar!/:?]
        at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:53) ~[arangodb-java-driver-5.0.3.jar!/:?]
        at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:49) ~[arangodb-java-driver-5.0.3.jar!/:?]
        at com.arangodb.internal.ArangoCollectionImpl.getDocument(ArangoCollectionImpl.java:134) ~[arangodb-java-driver-5.0.3.jar!/:?]
        at com.arangodb.springframework.core.template.ArangoTemplate.find(ArangoTemplate.java:510) ~[arangodb-spring-data-3.2.1.jar!/:?]
        at com.arangodb.springframework.core.template.ArangoTemplate.find(ArangoTemplate.java:520) ~[arangodb-spring-data-3.2.1.jar!/:?]
        at com.arangodb.springframework.repository.SimpleArangoRepository.findById(SimpleArangoRepository.java:122) ~[arangodb-spring-data-3.2.1.jar!/:?]
        at sun.reflect.GeneratedMethodAccessor26.invoke(Unknown Source) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_60]
        at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_60]
        at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:359) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:644) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
        at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:608) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
        at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
        at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
        at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
        at com.sun.proxy.$Proxy72.findById(Unknown Source) ~[?:?]
```

**Comment by Eremoue** [comment]
+1

**Comment by hanyu-natsu** [comment]
Any update on this?

**Comment by hanyu-natsu** [comment]
Any update on this ? We still have this problem

**Comment by fceller** [comment]
@hanyu-natsu have you tried 5.0.6? We did some overhaul in this area.

**Comment by hanyu-natsu** [comment]
@fceller thank you, I'll check out the new version.
Please close the issue. I'll reopen if it still happens.

**Comment by hanyu-natsu** [comment]
Please reopen this issue.
I'm still getting the same exception with java driver 5.0.7.

```
java.lang.NullPointerException: null
	at com.arangodb.internal.http.HttpCommunication.execute(HttpCommunication.java:87) ~[arangodb-java-driver-5.0.7.jar!/:?]
	at com.arangodb.internal.http.HttpProtocol.execute(HttpProtocol.java:49) ~[arangodb-java-driver-5.0.7.jar!/:?]
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:79) ~[arangodb-java-driver-5.0.7.jar!/:?]
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:59) ~[arangodb-java-driver-5.0.7.jar!/:?]
	at com.arangodb.internal.ArangoExecutorSync.execute(ArangoExecutorSync.java:55) ~[arangodb-java-driver-5.0.7.jar!/:?]
	at com.arangodb.internal.ArangoCollectionImpl.getDocument(ArangoCollectionImpl.java:129) ~[arangodb-java-driver-5.0.7.jar!/:?]
	at com.arangodb.springframework.core.template.ArangoTemplate.find(ArangoTemplate.java:510) ~[arangodb-spring-data-3.2.1.jar!/:?]
	at com.arangodb.springframework.core.template.ArangoTemplate.find(ArangoTemplate.java:520) ~[arangodb-spring-data-3.2.1.jar!/:?]
	at com.arangodb.springframework.repository.SimpleArangoRepository.findById(SimpleArangoRepository.java:122) ~[arangodb-spring-data-3.2.1.jar!/:?]
	at sun.reflect.GeneratedMethodAccessor111.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_191]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_191]
	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:359) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:644) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE
]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:608) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61) ~[spring-data-commons-2.1.3.RELEASE.jar!/:2.1.3.REL
EASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.3.RELEASE.jar!/:5.1.3.RELEASE]
	at com.sun.proxy.$Proxy125.findById(Unknown Source) ~[?:?]
```

**Comment by hanyu-natsu** [comment]
The issue occurs because in FallbackHostHandler, when failure iterations reach 3 it returns null as host. This results in NullPointerException in HttpCommunication.

I'm not pretty sure if incrementing iterations while failure is intended when I have only 1 host. At least it's possible to have a HostHandler to return null when getting host.

**Comment by hanyu-natsu** [comment]
Can you reopen this? @rashtao

**Comment by rashtao** [comment]
fixed here: https://github.com/arangodb/arangodb-java-driver/commit/1b66cac5e1a1596767bce6acd69a9dd7ea165bca?w=1

**Comment by rashtao** [comment]
fixed in v6.3.0
